src/g711/pcmacodec.cpp

```
@@ -25,31 +26,49 @@ PCMADecoder::~PCMADecoder()
 
 }
 
-int PCMAEncoder::Encode (SWORD *in,int inLen,BYTE* out,int outLen)
+AudioFrame::shared PCMAEncoder::Encode(const AudioBuffer::const_shared& audioBuffer)
 {
-	//Comprobamos las longitudes
-	if (outLen<inLen)
-		return 0;
-
-	//Y codificamos
+	const SWORD *in = audioBuffer->GetData();
+	int inLen = audioBuffer->GetNumSamples() * audioBuffer->GetNumChannels();
+	if(!in || inLen<=0) return nullptr;
 	for (int j = 0; j< inLen ;j++)
-		out[j]=linear2alaw(in[j]);
-
-	return inLen;
+	{
+		uint8_t alaw = linear2alaw(in[j]);
+		audioFrame->AppendMedia(&alaw, 1); 
+	}
+	audioFrame->SetLength(inLen);
+	return audioFrame;
 }
 
-int PCMADecoder::Decode(const AudioFrame::const_shared& audioFrame, SWORD* out,int outLen)
+
+int PCMADecoder::Decode(const AudioFrame::const_shared& audioFrame)
 {
 	//Comprobamos las longitudes
 	// inLen: compressed data length in bytes/samples one byte per sample for g711a, outLen: pcm data size in samples
-	int inLen = audioFrame->GetLength();
-	const uint8_t* in = audioFrame->GetData();
-	if (outLen<inLen || !in)
+	auto inLen = audioFrame->GetLength();
+	uint8_t* in = const_cast<uint8_t*>(audioFrame->GetData());
+	if (!in || !inLen)
 		return 0;
 
-	//Decodificamos
-	for (int j = 0; j< inLen ;j++)
-		out[j]=alaw2linear(in[j]);
-
-	return inLen;	
+	auto audioBuffer = std::make_shared<AudioBuffer>(inLen, 1);
+	for (int j = 0; j<inLen;j++)
+	{ 
+		if(audioBuffer->SetSampleAt(j, alaw2linear(*in)))
+			in++;
+		else	
+			return 0;
+	}
+	audioBufferQueue.push(std::move(audioBuffer));
+	return 1;	
 }
+
+AudioBuffer::shared PCMADecoder::GetDecodedAudioFrame()
+{
+	if(!audioBufferQueue.empty())
```

not big deal, but I preffer to fail early

```
if(audioBufferQueue.empty())
  return {};
auto audioBuffer = audioBufferQueue.front();
audioBufferQueue.pop();
return audioBuffer;
}

```